tico/utils/signature.py

```diff
@@ -0,0 +1,227 @@
+from typing import Sequence
+
+import numpy as np
+import torch
+from circle_schema import circle
+
+from tico.serialize.circle_mapping import to_circle_shape
+from tico.utils.dtype import circle_dtype_to_torch_dtype
+from tico.utils.installed_packages import is_dynamic_cache_available
+
+
+def isinstance_DynamicCache_safe(value):
+    if is_dynamic_cache_available():
+        from transformers.cache_utils import DynamicCache
+
+        return isinstance(value, DynamicCache)
+    else:
+        return False
+
+
+def flatten_and_convert_kwargs(kwargs: dict) -> dict[str, torch.Tensor]:
+    result = {}  # type: ignore[var-annotated]
+    for k, v in kwargs.items():
+        if v is None:
+            continue
+        elif isinstance(v, (list, tuple)):
+            # 1. handle list
+            def unpack_recursive(name, value, store=None):
+                if store is None:
+                    store = {}
+
+                if isinstance(value, (tuple, list)):
+                    for i, v in enumerate(value):
+                        # recursive call. Append index to name and explore lower level
+                        unpack_recursive(f"{name}_{i}", v, store)
+                else:
+                    # base type (scalar etc.) directly stored
+                    store[name] = value
+
+                return store
+
+            unpack_recursive(k, v, result)
+        elif isinstance_DynamicCache_safe(v):
+            # 2. handle DynamicCache
+            for idx, cache_val in enumerate(v.key_cache):
+                result[f"{k}_key_cache_{idx}"] = cache_val
+
+            for idx, cache_val in enumerate(v.value_cache):
+                result[f"{k}_value_cache_{idx}"] = cache_val
+        else:
+            result[k] = v
+
+    # 3. Convert to tensors
+    for k, v in result.items():
+        result[k] = v if isinstance(v, torch.Tensor) else torch.tensor(v)
+
+    return result
+
+
+def flatten_and_convert_args(args: Sequence) -> tuple:
```

Comment:
This flattening is required because `circle model` doesn't take tuples or list as argument. They must be flattened.

Reply:
Could you please show an example of flattening? Your comment is saying circle modle doesn't take tuples or list as argument. But the result of flatten is tuple. I don't understand. Maybe...

args = (Tensor1, Tensor2, scalar); is okay.
args = (Tensor1, tuple(Tensor21, Tensor22), scalar); is not okay.
So you want to make (Tensor1, newTensor2, scalar) by flattening. Is it right?

I look up the flattening example in test_signature.py. But I can't find.

Reply:
@glistening I added the test test_bind_tuple, PTAL!